## Part 1: Distributed Data Processing with Spark

### Task 1.1: Spark Environment Setup & Data Loading

In this task, SparkSession was configured for distributed data processing and load the NYC Yellow Taxi dataset into a Spark DataFrame.

We have completed the following:
- Configuration of Spark with appropriate memory settings and adaptive query execution
- Loading the dataset from a Parquet file
- Inspecting the schema to verify correct data types
- Calulating the total row count and partition count
- Comparation between Spark and Pandas loading performance

These steps enables the foundation for scalable and distributed data processing and highlighting the advantages of Spark's distributed processing  for large datasets.

In [ ]:
from pyspark.sql import SparkSession

spark = (
    SparkSession
    .builder
    .appName("Sparky")
    .master("local[*]")
    .config("spark..executor.memory", "6g")# excuting the task assigned,
    .config("spark.driver.memory", "8g") # breaking up the task and distirbuting it ,colab has 12.7
    .config("spark.sql.adaptive.enabled","true")#an optimization technique in Spark SQL that makes use of the runtime statistics to choose the most efficient query execution plan
    .config("spark.sql.legacy.parquet.nanosAsLong", "true")#parquet file was created using pandas in pervious assignments, there is an incompatablity issue, throws an illegal type error without this
    .getOrCreate()

)

#print(spark)
spark

In [ ]:
import requests
import os
try:
    file = "yellow_taxi_data.parquet"
    if not os.path.exists(file):
        print("the taxi data file is missing,attempting to download")
        local_filename = "yellow_taxi_data.parquet"
        url1 ="https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2024-01.parquet"
        with requests.get(url1, stream=True) as r:
           r.raise_for_status()
           with open(local_filename, 'wb') as f:
              for chunk in r.iter_content(chunk_size=8192):
                  f.write(chunk)
        print("yelllow completed")
except Exception as e:
    print("this error occured downloading the parquet file: ", e)


In [ ]:
#loading data into spark df
taxi = spark.read.parquet("yellow_taxi_data.parquet")

required_col = ["tpep_pickup_datetime","tpep_dropoff_datetime","PULocationID","DOLocationID","passenger_count","trip_distance","fare_amount","fare_amount","total_amount","payment_type"]
datetime_field =["tpep_pickup_datetime","tpep_dropoff_datetime"]
#compares columns in df to columns in required col
contains_col =all(column in taxi.columns for column in required_col)# true for all the condition( if column in the dataset is in the required field list)
if contains_col:
     print ("dataset has all required columns")
else:
      missing_col =all(column not in taxi.columns for column in required_col)#true for all the condition(if column in the dataset is not in the required field)
      print ("dataset has is missing required columns",missing_col)

      datetime_col= taxi.select(datetime_field).schema
      #print(datetime_col)
      if all(datetime_col[column] == taxi.Datetime for column in datetime_field):#true for all condition(if column in the datetime_list ,is in the dataset and is of type datetime )
         print ("dataset has all required datetime columns")
      else:
         print ("dataset MISSING required datetime columns")

#displaying schema
print(taxi)
taxi.printSchema()
taxi.dtypes

In [ ]:
print("the total number of rows in taxi ",taxi.count())
#partition count
print("the number of partitions ", taxi.rdd.getNumPartitions()) #rrd - Resilient Distributed Dataset - gives you low-level control, great for handling custom formats. But it’s more verbose, harder to optimize, and less friendly


In [ ]:
import time
import pandas as pd
def measure_time(func):
 start = time.time()
 result = func()
 end = time.time()
 del result
 return  end - start

pd_load_time = measure_time(lambda:pd.read_parquet("yellow_taxi_data.parquet"))
sparkload_time = measure_time(lambda:spark.read.parquet("yellow_taxi_data.parquet"))
spark_speedup_percent = (sparkload_time/pd_load_time)*100

print("the load time for pandas : ",pd_load_time,"\nthe load time for spark : ",sparkload_time, "\nresulting in a ",spark_speedup_percent," percent decrease for loading data")



the load time for pandas :  1.1427323818206787 
the load time for spark :  0.3038012981414795 
resulting in a  26.58551581932444  percent decrease for loading data


Spark loads the Parquet file significantly faster than Pandas because of its distributed architecture. Spark reads the data in parallel across multiple cores partitions, whereas Pandas
processes the file sequentially in a single thread. Furthermore, Spark evaluates operations lazily, meaning it only reads the necessary metadata until an action is explicitly called.

### Task 1.2: Data Cleaning & Feature Engineering

The taxi dataset may contain missing or invalid records. In this section, cleaning the dataset and engineering new features for analysis.

Data cleaning included:
- Removing rows with null values in critical columns
- Filtering invalid trips (such as negative fares, zero distance, outliers)

Feature engineering includes:
- Trip duration
- Trip speed
- Pickup hour
- Day of week
- Tip percentage

the tracking of how many rows are removed at each step to understand data quality, ensuring the dataset is accurate, consistent, and suitable for valid analytics.

In [ ]:
try:

  total= taxi.count()
  total_row = taxi.count()

  #dropping null
  taxi = taxi.dropna(subset=["tpep_pickup_datetime","tpep_dropoff_datetime","PULocationID","DOLocationID","fare_amount","trip_distance"])
  print("the number of null rows dropped : ",total_row - taxi.count())
  total_row = taxi.count()

  #filtering
  # zero/negative distance
  taxi = taxi.filter(taxi.trip_distance > 0)
  print("the number of rows dropped with zero/negative distance: ",total_row - taxi.count())

  # negative fares
  total_row = taxi.count()
  taxi =taxi.filter((taxi.fare_amount > -1) & (taxi.fare_amount <500))
  print("the number of rows dropped with negative fares  : ",total_row - taxi.count())

  #dropoff before pickup
  total_row = taxi.count()
  taxi =taxi.filter(taxi.tpep_dropoff_datetime > taxi.tpep_pickup_datetime)
  print("the number of rows dropped with dropoff before pickup : ",total_row - taxi.count())

  #total rows dropped
  print("totals row dropped ", total-taxi.count())
  print("total remaining rows: ", taxi.count())
except Exception as e:
    print("Cleaning step failed:", e)


the number of null rows dropped :  0
the number of rows dropped with zero/negative distance:  60371
the number of rows dropped with negative fares  :  33601
the number of rows dropped with dropoff before pickup :  112
totals row dropped  94084
total remaining rows:  2870540


 derived columns using Spark functions:



In [ ]:
import pyspark.sql.functions as sf
try:
  new_taxi = taxi.withColumn('trip_duration_minutes',taxi.tpep_dropoff_datetime - taxi.tpep_pickup_datetime)

  new_taxi = new_taxi.withColumn('trip_speed_mph',sf.try_divide(new_taxi.trip_distance, (new_taxi.trip_duration_minutes).cast("long"))/60)#fix has neg

  new_taxi = new_taxi.withColumn('pickup_hour',sf.hour(new_taxi.tpep_pickup_datetime))

  new_taxi = new_taxi.withColumn('pickup_day_of_week',sf.dayofweek(new_taxi.tpep_pickup_datetime))

  new_taxi = new_taxi.withColumn('tip_percentage',sf.try_divide(new_taxi.tip_amount,new_taxi.fare_amount)*100)
  print("scccessfully created new features")
  new_taxi.printSchema()
except Exception as e:
    print("feature engineering failed: ",e)



scccessfully created new features
root
 |-- VendorID: integer (nullable = true)
 |-- tpep_pickup_datetime: timestamp_ntz (nullable = true)
 |-- tpep_dropoff_datetime: timestamp_ntz (nullable = true)
 |-- passenger_count: long (nullable = true)
 |-- trip_distance: double (nullable = true)
 |-- RatecodeID: long (nullable = true)
 |-- store_and_fwd_flag: string (nullable = true)
 |-- PULocationID: integer (nullable = true)
 |-- DOLocationID: integer (nullable = true)
 |-- payment_type: long (nullable = true)
 |-- fare_amount: double (nullable = true)
 |-- extra: double (nullable = true)
 |-- mta_tax: double (nullable = true)
 |-- tip_amount: double (nullable = true)
 |-- tolls_amount: double (nullable = true)
 |-- improvement_surcharge: double (nullable = true)
 |-- total_amount: double (nullable = true)
 |-- congestion_surcharge: double (nullable = true)
 |-- Airport_fee: double (nullable = true)
 |-- trip_duration_minutes: interval day to second (nullable = true)
 |-- trip_speed_mph: do

### Task 1.3: Spark SQL Analytics

In this section, using Spark SQL to perform analytical queries on the cleaned dataset.

 registering the dataframe as a temporary SQL view including:
- busiest taxi demand periods
- Trip efficiency across days
- Revenue distribution by location
- Cumulative trip patterns
- Comparison of trip categories

Each query result is has a short interpretation.

In [ ]:
#creating temporary view
new_taxi.createOrReplaceTempView("data")
no_cache=[]

In [ ]:
try:
  start =time.time()
  query1 = spark.sql('''SELECT pickup_hour,
      COUNT(*) AS trip_count,
      AVG(fare_amount) AS avg_fare,
      AVG(tip_percentage) AS avg_tip_pct
  FROM data
  GROUP BY pickup_hour
  ORDER BY trip_count DESC
  LIMIT 10''')
  no_cache.append(time.time() - start)
  query1.show()
except Exception as e:
    print("query failed: ",e)

+-----------+----------+------------------+------------------+
|pickup_hour|trip_count|          avg_fare|       avg_tip_pct|
+-----------+----------+------------------+------------------+
|         18|    206346|17.007660240566636|22.776572903640027|
|         17|    200341|18.114963587084002|22.339846207820127|
|         16|    184985|19.455140308673965| 21.83536272199866|
|         15|    184022|19.108745041353632|19.799061000856273|
|         19|    178849| 17.62270725584169|22.854482751189064|
|         14|    178051|19.267840337880916| 19.79511517843723|
|         13|    165361|18.417337824517265|  19.7851853540971|
|         12|    159931|17.793507637669034| 19.74033458023976|
|         21|    155940|18.285690265486643|21.877142260190507|
|         20|    155579| 18.04789496011668|  22.1687056636625|
+-----------+----------+------------------+------------------+



The 10 busiest pickup hours range from12pm to 9pm indicating that most pickups are a contribution of employess departing the workplace for lunch and the two busiest hours 5pm.and 6pm are employees departing the workplace which also had the second and third highest tip percentage implying passenger tip more when they are heading home

In [ ]:
try:
  start =time.time()
  query2 = spark.sql('''SELECT pickup_day_of_week,
      AVG(trip_speed_mph) AS avg_speed_mph,
      AVG(trip_distance) AS avg_distance,
      AVG(trip_duration_minutes) AS avg_duration
  FROM data
  GROUP BY pickup_day_of_week
  ORDER BY avg_speed_mph DESC
  LIMIT 1''')
  no_cache.append(time.time() - start)
  query2.show()
except Exception as e:
    print("query failed: ",e)

+------------------+--------------------+-----------------+--------------------+
|pickup_day_of_week|       avg_speed_mph|     avg_distance|        avg_duration|
+------------------+--------------------+-----------------+--------------------+
|                 3|8.080728559873352E-5|4.245027468613659|INTERVAL '0 00:16...|
+------------------+--------------------+-----------------+--------------------+



thursday had the highest average speed, with an average speed of 8mph and an average distance of 4miles for average duration of 16seconds

In [ ]:
try:
 start =time.time()
 query3 = spark.sql('''WITH revenue_per_day AS (
      SELECT
          pickup_day_of_week,
          PULocationID,
          SUM(fare_amount) AS total_revenue
      FROM data
      GROUP BY pickup_day_of_week, PULocationID
  )
  SELECT *
  FROM (
      SELECT
          pickup_day_of_week,
          PULocationID,
          total_revenue,
          RANK() OVER (
              PARTITION BY pickup_day_of_week
              ORDER BY total_revenue DESC
          ) AS rank
      FROM revenue_per_day
  ) ranked
  WHERE rank <= 5''')
 no_cache.append(time.time() - start)
 query3.show()
except Exception as e:
    print("query failed: ",e)

+------------------+------------+------------------+----+
|pickup_day_of_week|PULocationID|     total_revenue|rank|
+------------------+------------+------------------+----+
|                 1|         132|        1221514.95|   1|
|                 1|         138|487599.09999999875|   2|
|                 1|         230|234661.10000000018|   3|
|                 1|         186| 179522.8999999999|   4|
|                 1|          79|173510.82000000012|   5|
|                 2|         132| 1593427.760000003|   1|
|                 2|         138| 648123.5200000003|   2|
|                 2|         161|299011.26000000024|   3|
|                 2|         186|236408.26000000077|   4|
|                 2|         236| 236287.6000000009|   5|
|                 3|         132|1393573.0899999975|   1|
|                 3|         138| 597059.8199999982|   2|
|                 3|         161| 398150.4399999998|   3|
|                 3|         236| 307749.3599999984|   4|
|             

Location id 132 consistently ranks as the highest revenue erraning pickup location across almost all days of the week indicating that it represents a infustructure based building such as an airport or a highly commercial area which in New York city many builds are skyscrapers due to limited estate.

In [ ]:
try:
  start =time.time()
  query4 = spark.sql('''WITH hourly_counts AS (
      SELECT
          pickup_hour,
          COUNT(*) AS trip_count
      FROM data
      GROUP BY pickup_hour
  ),
  cumulative AS (
      SELECT
          pickup_hour,
          trip_count,
          SUM(trip_count) OVER (ORDER BY pickup_hour) AS cumulative_trips,
          SUM(trip_count) OVER () AS total_trips
      FROM hourly_counts
  )
  SELECT *
  FROM cumulative
  WHERE cumulative_trips >= 0.5 * total_trips
  ORDER BY pickup_hour
  LIMIT 1''')
  no_cache.append(time.time() - start)
  query4.show()
except Exception as e:
    print("query failed: ",e)

+-----------+----------+----------------+-----------+
|pickup_hour|trip_count|cumulative_trips|total_trips|
+-----------+----------+----------------+-----------+
|         15|    184022|         1545539|    2870540|
+-----------+----------+----------------+-----------+



By 3pm the cumulative number of trips surpassed 50% of the total daily volume. This confirms that the majority of taxi demand is concentrated in the morning and midday hours.

In [ ]:
try:
  start =time.time()
  query5 =spark.sql('''SELECT
      CASE
          WHEN trip_distance < 2 THEN 'Short'
          WHEN trip_distance BETWEEN 2 AND 10 THEN 'Medium'
          ELSE 'Long'
      END AS trip_category,
      AVG(fare_amount) AS avg_fare,
      AVG(trip_distance) AS avg_distance,
      AVG(tip_percentage) AS avg_tip_pct
  FROM data
  GROUP BY trip_category
  ORDER BY avg_tip_pct DESC''')
  no_cache.append(time.time() - start)
  query5.show()

  no_cache.append(sum(no_cache)/len(no_cache))
except Exception as e:
    print("query failed: ",e)


+-------------+------------------+------------------+------------------+
|trip_category|          avg_fare|      avg_distance|       avg_tip_pct|
+-------------+------------------+------------------+------------------+
|        Short| 9.910277383915167|1.1298463557821081|23.071051197958855|
|         Long| 64.64381167971426| 21.69866275091764| 21.93473956256546|
|       Medium|22.167563119139462|3.9614304445589337|18.561826661812383|
+-------------+------------------+------------------+------------------+



long trips had the highest average fare of 64.00 but a average tip percentages of 21%. Medium trip had the lowest tip percentage of 18% with a average fare of $22.00 . The short trips had the highest tip percentage of 23.00 to a average fare of 10.00. It is expected that short trips have a high tip percentage as due to low fare, a small tip is of a high percentage. But unexpectedly long trips tend to have lowest tip percentage due to highest average fare and passenger fatigue as they tend to give generally smaller tips

### Task 1.4: Performance Optimization

In this section exploring techniques to improve  performance by :

- Cache the cleaned dataframe and compare query time before and after caching

- write the dataset to Parquet format partitioned by pickup hour
- demonstrating partition pruning by reading a single partition
- Analyze the execution plan of a query using explain method

These optimizations demonstrate how Spark can efficiently process large datasets.


In [ ]:
try:
  new_taxi.cache()
  cached = []

  start =time.time()
  query1 = spark.sql('''SELECT pickup_hour,
      COUNT(*) AS trip_count,
      AVG(fare_amount) AS avg_fare,
      AVG(tip_percentage) AS avg_tip_pct
    FROM data
    GROUP BY pickup_hour
    ORDER BY trip_count DESC
    LIMIT 10''')
  cached.append(time.time() - start)

  start =time.time()
  query2 = spark.sql('''SELECT pickup_day_of_week,
      AVG(trip_speed_mph) AS avg_speed_mph,
      AVG(trip_distance) AS avg_distance,
      AVG(trip_duration_minutes) AS avg_duration
    FROM data
    GROUP BY pickup_day_of_week
    ORDER BY avg_speed_mph DESC
    LIMIT 1''')
  cached.append(time.time() - start)

  start =time.time()
  query3 = spark.sql('''WITH revenue_per_day AS (
      SELECT
          pickup_day_of_week,
          PULocationID,
          SUM(fare_amount) AS total_revenue
      FROM data
      GROUP BY pickup_day_of_week, PULocationID
    )
    SELECT *
    FROM (
      SELECT
          pickup_day_of_week,
          PULocationID,
          total_revenue,
          RANK() OVER (
              PARTITION BY pickup_day_of_week
              ORDER BY total_revenue DESC
          ) AS rank
      FROM revenue_per_day
    ) ranked
    WHERE rank <= 5''')
  cached.append(time.time() - start)

  start =time.time()
  query4 = spark.sql('''WITH hourly_counts AS (
      SELECT
          pickup_hour,
          COUNT(*) AS trip_count
      FROM data
      GROUP BY pickup_hour
    ),
    cumulative AS (
      SELECT
          pickup_hour,
          trip_count,
          SUM(trip_count) OVER (ORDER BY pickup_hour) AS cumulative_trips,
          SUM(trip_count) OVER () AS total_trips
      FROM hourly_counts
    )
    SELECT *
    FROM cumulative
    WHERE cumulative_trips >= 0.5 * total_trips
    ORDER BY pickup_hour
    LIMIT 1''')
  cached.append(time.time() - start)
  #query4.show()

  start =time.time()
  query5 =spark.sql('''SELECT
      CASE
          WHEN trip_distance < 2 THEN 'Short'
          WHEN trip_distance BETWEEN 2 AND 10 THEN 'Medium'
          ELSE 'Long'
      END AS trip_category,
      AVG(fare_amount) AS avg_fare,
      AVG(trip_distance) AS avg_distance,
      AVG(tip_percentage) AS avg_tip_pct
  FROM data
  GROUP BY trip_category
  ORDER BY avg_tip_pct DESC''')
  cached.append(time.time() - start)
  #query5.show()

  cached.append(sum(cached) / len(cached))

  print("times for querys with no caching",no_cache)
  print("times for querys with caching",cached)

  speedup = no_cache[-1]/cached[-1]

  for i in range(len(no_cache)):
    speedup =   no_cache[i]/cached[i]*100
    print("the caching speedup for query :" ,i," ",speedup )
except Exception as e:
    print("optimization comparaison failed: ",e)

times for querys with no caching [0.012937784194946289, 0.02021956443786621, 0.02000713348388672, 0.0393068790435791, 0.02431344985961914, 0.02335696220397949]
times for querys with caching [0.01835322380065918, 0.011260986328125, 0.01697063446044922, 0.03001570701599121, 0.013747215270996094, 0.01806955337524414]
the caching speedup for query : 0   70.49325140622767
the caching speedup for query : 1   179.55411585365854
the caching speedup for query : 2   117.89266647934812
the caching speedup for query : 3   130.95436673418325
the caching speedup for query : 4   176.86090877558098
the caching speedup for query : 5   129.2614251107018


Spark caching provided a minute speedup for query 1 and query 4 but not for the rest while these times were of 0.01 second difference and query1 and query 4 both used the pickup hours while the rest of queries had to intersecting data to retrieve but any queries after using the same data my benefit from a speed increase. however this varies and is subjected to change

In [ ]:
try:
  new_taxi.write.partitionBy("pickup_hour").mode("overwrite").parquet("hours.parquet")
  hour_taxi = spark.read.parquet("hours.parquet")
  #hour_taxi.show()
  h17 = hour_taxi.filter(hour_taxi.pickup_hour == 16)
  h17.show()
  print("the number of partitions ", hour_taxi.rdd.getNumPartitions()) #rrd - Resilient Distributed Dataset - gives you low-level control, great for handling custom formats. But it’s more verbose, harder to optimize, and less friendly

  print("the following is the partition pruning ")
  h17.explain(True)

  print("the following is the complex sql query plan")
  query3.explain()
except Exception as e:
    print("Partitioning  failed:", e)

In the partitions plan the Parsed Logical Plan, spark converts the query filter by pickup hour 16 into a logical tree with each coloumn given a unqiue id without checking the dataset. in the Analyzed Logical Plan spark checks the dataset, establishing data types and names. in the
Optimized Logical Plan  checks for only where pickup equals 16. in Physical Plan spark reads the parquet using partition filters.

The execution plan for query 3 involves several key operations a hashaggregate is used to group the data by day and location to calculate total revenue. This is followed by an exchange operation, which shuffles the data across partitions across the cluster to prepare for the window function that ranks the revenues.

# Part 2: RAG Pipeline over Transportation Documents


### Task 2.1: Document Collection & Ingestion

In this section,collecting and processing documents related to NYC transportation policy by :
- Curating 5–10 relevant documents using UWI linC
- Extracting text using PDF loaders
- Calculating total pages and character counts
The documents are stored in the docs/ directory.These documents forms the knowledge base for the RAG system.

In [ ]:
!pip install pypdf langchain langchain-community chromadb openai sentence-transformers matplotlib numpy

In [ ]:
from pypdf import PdfReader
import os
from langchain_community.document_loaders import PyPDFDirectoryLoader
try:
  loader = PyPDFDirectoryLoader("/content/docs")
  raw_documents = loader.load()
  print(f"Loaded {len(raw_documents)} pages from all PDFs")
  print(f"First document metadata: {raw_documents[0].metadata}")
  #print(f"First 200 chars: {raw_documents[0].page_content[:200]}")

  # Check for empty or very short pages
  for doc in raw_documents:
    if len(doc.page_content.strip()) < 50:
       print(f"Short/empty page: {doc.metadata}")
  # Check character count distribution
  lengths = [len(d.page_content) for d in raw_documents]
  print(f"tatol chars per page: {sum(lengths):.0f}")
  print(f"Avg chars per page: {sum(lengths)/len(lengths):.0f}")
  print(f"Min: {min(lengths)}, Max: {max(lengths)}")
except Exception as e:
    print("document loading failed:",e)

Loaded 157 pages from all PDFs
First document metadata: {'producer': 'Acrobat Distiller 11.0 (Windows)', 'creator': 'Arbortext Advanced Print Publisher 9.0.114/W', 'creationdate': '2019-04-19T05:40:40-04:00', 'moddate': '2019-04-19T05:40:40-04:00', 'title': 'IRS798431 281..306', 'source': '/content/documents/EBSCO-FullText-03_21_2026 (2).pdf', 'total_pages': 27, 'page': 0, 'page_label': '1'}
tatol chars per page: 441601
Avg chars per page: 2813
Min: 77, Max: 7829



During data ingestion, a total of 157 pages were extracted with the average page having 2813 characters which the lowest character count was 77 and the highest character count was 7829

### Task 2.2: Chunking & Embedding

To enable efficient retrieval spliting documents into smaller chunks and convert into vector embeddings by:
- using RecursiveCharacterTextSplitter (chunk_size=1000, overlap=200)
- analyze chunk size distribution
- generating embeddings using "all-MiniLM-L6-v2"
- storeing embeddings in ChromaDB

This step is critical for balancing context size and retrieval accuracy in the RAG pipeline.

In [ ]:
from langchain_text_splitters import RecursiveCharacterTextSplitter

text_splitter1000 = RecursiveCharacterTextSplitter(
chunk_size=1000,
chunk_overlap=200,
separators=["\n\n", "\n", ". ", " ", ""])
chunks1000 = text_splitter1000.split_documents(raw_documents)
print(f"Split {len(raw_documents)} pages into {len(chunks1000)} chunks")
try:

  chunk_a = chunks1000[0].page_content
  chunk_b = chunks1000[1].page_content

  overlap_len = 0
  for length in range(min(300, len(chunk_a)), 0, -1):
    if chunk_b.startswith(chunk_a[-length:]):
      overlap_len = length
      break
  print(f"Overlap length: {overlap_len} chars")
  if overlap_len > 0:
    print(f"Overlap text: {chunk_a[-overlap_len:][:100]}...")
  else:
    print("No overlap found (chunks may be from different documents)")
except Exception as e:
    print("overlap anaylsis failed:", e)


Split 157 pages into 600 chunks
Overlap length: 167 chars
Overlap text: theatergoers as trips where the drop-off or pickup is near Broadway within thirty
minutes of the beg...


In [ ]:

import matplotlib.pyplot as plt
for i, chunk in enumerate(chunks1000[:3]):
    print(f"\n=== Chunk {i} ===")
    print(f"Length: {len(chunk.page_content)} chars")
    print(f"Source: {chunk.metadata}")
    print(f"Preview: {chunk.page_content[:200]}...")


'''sizes = [len(c.page_content) for c in chunks1000]
plt.figure(figsize=(8, 4))
plt.hist(sizes, bins=30, edgecolor='black')
plt.xlabel('Chunk Size (characters)')
plt.ylabel('Count')
plt.title('Distribution of Chunk Sizes')
plt.tight_layout()
plt.show()'''



=== Chunk 0 ===
Length: 928 chars
Source: {'producer': 'Acrobat Distiller 11.0 (Windows)', 'creator': 'Arbortext Advanced Print Publisher 9.0.114/W', 'creationdate': '2019-04-19T05:40:40-04:00', 'moddate': '2019-04-19T05:40:40-04:00', 'title': 'IRS798431 281..306', 'source': '/content/documents/EBSCO-FullText-03_21_2026 (2).pdf', 'total_pages': 27, 'page': 0, 'page_label': '1'}
Preview: Original Article
Do Tourists Tip More
Than Local Consumers?
Evidence from Taxi Rides
in New York City
Amir B. Ferreira Neto 1, Adam Nowak 1
and Amanda Ross 2
Abstract
Given the resurgence of cities as...

=== Chunk 1 ===
Length: 995 chars
Source: {'producer': 'Acrobat Distiller 11.0 (Windows)', 'creator': 'Arbortext Advanced Print Publisher 9.0.114/W', 'creationdate': '2019-04-19T05:40:40-04:00', 'moddate': '2019-04-19T05:40:40-04:00', 'title': 'IRS798431 281..306', 'source': '/content/documents/EBSCO-FullText-03_21_2026 (2).pdf', 'total_pages': 27, 'page': 0, 'page_label': '1'}
Preview: theatergoers a

"sizes = [len(c.page_content) for c in chunks1000]\nplt.figure(figsize=(8, 4))\nplt.hist(sizes, bins=30, edgecolor='black')\nplt.xlabel('Chunk Size (characters)')\nplt.ylabel('Count')\nplt.title('Distribution of Chunk Sizes')\nplt.tight_layout()\nplt.show()"

**Experiment**

In [ ]:

text_splitter500 = RecursiveCharacterTextSplitter(
chunk_size=500,
chunk_overlap=200,
separators=["\n\n", "\n", ". ", " ", ""])
chunks500 = text_splitter500.split_documents(raw_documents)
print(f"Split {len(raw_documents)} pages into {len(chunks500)} chunks")

try:

  chunk_a = chunks500[0].page_content
  chunk_b = chunks500[1].page_content

  overlap_len = 0
  for length in range(min(300, len(chunk_a)), 0, -1):
    if chunk_b.startswith(chunk_a[-length:]):
      overlap_len = length
      break
  print(f"Overlap length: {overlap_len} chars")
  if overlap_len > 0:
    print(f"Overlap text: {chunk_a[-overlap_len:][:100]}...")
  else:
    print("No overlap found (chunks may be from different documents)")
except Exception as e:
    print("overlap anaylsis failed:", e)

Split 157 pages into 1478 chunks
Overlap length: 162 chars
Overlap text: City. Taxi service is an endogenous service; however, taxis also contribute to the
demand and provis...


In [ ]:

text_splitter2000 = RecursiveCharacterTextSplitter(
chunk_size=2000, # Target characters per chunk
chunk_overlap=200, # Overlap between consecutive chunks
separators=["\n\n", "\n", ". ", " ", ""])
chunks2000 = text_splitter2000.split_documents(raw_documents)
print(f"Split {len(raw_documents)} pages into {len(chunks2000)} chunks")
try:
  chunk_a = chunks2000[0].page_content
  chunk_b = chunks2000[1].page_content
  overlap_len = 0
  for length in range(min(300, len(chunk_a)), 0, -1):
    if chunk_b.startswith(chunk_a[-length:]):
      overlap_len = length
      break
  print(f"Overlap length: {overlap_len} chars")
  if overlap_len > 0:
    print(f"Overlap text: {chunk_a[-overlap_len:][:100]}...")
  else:
    print("No overlap found (chunks may be from different documents)")
except Exception as e:
    print("overlap anaylsis failed:", e)

Split 157 pages into 310 chunks
Overlap length: 0 chars
No overlap found (chunks may be from different documents)


In [ ]:
from langchain_community.embeddings import HuggingFaceEmbeddings
try:
  embedding_model = HuggingFaceEmbeddings(
  model_name="all-MiniLM-L6-v2"
  )

  test_embedding = embedding_model.embed_query("What are the rules for zero-emission taxi fleets?")
  print(f"Embedding dimension: {len(test_embedding)}")
  print(f"First  values: {test_embedding[:3]}")
except Exception as e:
    print("Embedding failed:", e)


Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


Embedding dimension: 384
First  values: [0.0795130804181099, -0.018346719443798065, 0.011498321779072285]


In [ ]:
from langchain_community.vectorstores import Chroma
try:
    vectorstore = Chroma.from_documents(
    documents=chunks1000,
    embedding=embedding_model,
    persist_directory="./chroma_db",
    collection_name="assignment3"
    )
    print(f"Indexed {len(chunks1000)} chunks in ChromaDB")
    query = "What are the main topics covered in this document?"
    results = vectorstore.similarity_search(query, k=3)
    for i, doc in enumerate(results):
      print(f"\n--- Result {i+1} (source: {doc.metadata.get('source', 'unknown')}) ---")
      print(doc.page_content[:300])
except Exception as e:
    print("storing embedding using chromaDB failed:", e)


    """results_with_scores = vectorstore.similarity_search_with_score(query, k=5)
for doc, score in results_with_scores:
 print(f"Score: {score:.4f} | Source: {doc.metadata.get('source', '?')} "
 f"| Preview: {doc.page_content[:80]}...")"""


Indexed 600 chunks in ChromaDB

--- Result 1 (source: /content/documents/AlonsoMora-Ondemandhighcapacityridesharing-2017.pdf) ---
This content downloaded from 
64.28.139.201 on Sat, 21 Mar 2026 17:34:57 UTC 
All use subject to https://about.jstor.org/terms

--- Result 2 (source: /content/documents/AlonsoMora-Ondemandhighcapacityridesharing-2017.pdf) ---
This content downloaded from 
64.28.139.201 on Sat, 21 Mar 2026 17:34:57 UTC 
All use subject to https://about.jstor.org/terms

--- Result 3 (source: /content/documents/AlonsoMora-Ondemandhighcapacityridesharing-2017.pdf) ---
This content downloaded from 
64.28.139.201 on Sat, 21 Mar 2026 17:34:57 UTC 
All use subject to https://about.jstor.org/terms


In [ ]:
 # create the vector store from chunks
vectorstore500 = Chroma.from_documents(
documents=chunks500,
embedding=embedding_model,
persist_directory="./chroma_db",
collection_name="assignment3-c500")
print(f"Indexed {len(chunks500)} chunks in ChromaDB")
query = "What are the main topics covered in this document?"
results500 = vectorstore500.similarity_search(query, k=3)
for i, doc in enumerate(results500):
  print(f"\n--- Result {i+1} (source: {doc.metadata.get('source', 'unknown')}) ---")
  print(doc.page_content[:300])


Indexed 1478 chunks in ChromaDB

--- Result 1 (source: /content/documents/AlonsoMora-Ondemandhighcapacityridesharing-2017.pdf) ---
access to Proceedings of the National Academy of Sciences of the United States of America
This content downloaded from 
64.28.139.201 on Sat, 21 Mar 2026 17:34:57 UTC 
All use subject to https://about.jstor.org/terms

--- Result 2 (source: /content/documents/AlonsoMora-Ondemandhighcapacityridesharing-2017.pdf) ---
access to Proceedings of the National Academy of Sciences of the United States of America
This content downloaded from 
64.28.139.201 on Sat, 21 Mar 2026 17:34:57 UTC 
All use subject to https://about.jstor.org/terms

--- Result 3 (source: /content/documents/AlonsoMora-Ondemandhighcapacityridesharing-2017.pdf) ---
access to Proceedings of the National Academy of Sciences of the United States of America
This content downloaded from 
64.28.139.201 on Sat, 21 Mar 2026 17:34:57 UTC

In [ ]:
vectorstore2000 = Chroma.from_documents(
documents=chunks2000,
embedding=embedding_model,
persist_directory="./chroma_db",
collection_name="assignment3-c2000")
print(f"Indexed {len(chunks500)} chunks in ChromaDB")
query = "What are the main topics covered in this document?"
results2000 = vectorstore2000.similarity_search(query, k=3)
for i, doc in enumerate(results2000):
  print(f"\n--- Result {i+1} (source: {doc.metadata.get('source', 'unknown')}) ---")
  print(doc.page_content[:300])


Indexed 1478 chunks in ChromaDB

--- Result 1 (source: /content/documents/AlonsoMora-Ondemandhighcapacityridesharing-2017.pdf) ---
set of constraintsZ, and allows for multiple passengers per vehi-
cle. A passenger is a past request that has been picked up by a
vehicle and that is now en route to its destination. We denote by
Pv the set of passengers for vehicle v∈V . In a second step, the
method also allows to rebalance the flee

--- Result 2 (source: /content/documents/AlonsoMora-Ondemandhighcapacityridesharing-2017.pdf) ---
set of constraintsZ, and allows for multiple passengers per vehi-
cle. A passenger is a past request that has been picked up by a
vehicle and that is now en route to its destination. We denote by
Pv the set of passengers for vehicle v∈V . In a second step, the
method also allows to rebalance the flee

--- Result 3 (source: /content/documents/AlonsoMora-Ondemandhighcapacityridesharing-2017.pdf) ---
set of constraintsZ, and allows for multiple passengers per vehi-
cle

Chunk Size Anal[link text](https:// [link text](https://))ysis:**
Comparing chunk sizes of 500, 1000, and 2000 characters reveals that **1000 characters** provides the best balance. The 500-character chunks often lack sufficient context, splitting sentences abruptly and causing the LLM to miss nuanced details. The 2000-character chunks capture full context but introduce too much noise, occasionally confusing the retriever with irrelevant surrounding text. The 1000-character chunks consistently retrieved the most focused and accurate context.

### Task 2.3: RAG Pipeline Implementation

In this section implementing a complete Retrieval-Augmented Generation (RAG) pipeline.

We test the pipeline using multiple transportation-related questions and display:
- Generated answers
- Retrieved sources
- Context chunks

In [ ]:
try:
  retriever = vectorstore.as_retriever(search_kwargs={"k": 4})
  # Test retrieval
  retrieved_docs = status =.invoke("What are the main topics covered in this document?")############
  print(f"Retrieved {len(retrieved_docs)} chunks")
  for doc in retrieved_docs:
    print(f" - {doc.metadata.get('source', '?')} | {doc.page_content[:100]}...")
except Exception as e:
    print("retrieval failed:",e)

Retrieved 4 chunks
 - /content/documents/AlonsoMora-Ondemandhighcapacityridesharing-2017.pdf | This content downloaded from 
64.28.139.201 on Sat, 21 Mar 2026 17:34:57 UTC...
 - /content/documents/AlonsoMora-Ondemandhighcapacityridesharing-2017.pdf | This content downloaded from 
64.28.139.201 on Sat, 21 Mar 2026 17:34:57 UTC...
 - /content/documents/AlonsoMora-Ondemandhighcapacityridesharing-2017.pdf | This content downloaded from 
64.28.139.201 on Sat, 21 Mar 2026 17:34:57 UTC...
 - /content/documents/AlonsoMora-Ondemandhighcapacityridesharing-2017.pdf | This content downloaded from 
64.28.139.201 on Sat, 21 Mar 2026 17:34:57 UTC...


in this case all the chunks retrieved the correct documents, not affecting the retriever

In [ ]:
try:
  def format_context(docs):
 # """Format retrieved documents into a numbered context string."""
   context_parts = []
   for i, doc in enumerate(docs, 1):
    source = doc.metadata.get("source", "Unknown")
    page = doc.metadata.get("page", "?")
    context_parts.append(f"[Source {i}: {source}, Page {page}]\n{doc.page_content}")
    return "\n\n---\n\n".join(context_parts)
  # Test
  context = format_context(retrieved_docs)
  print(context[:500])
except Exception as e:
    print("function format_context failed:",e)

[Source 1: /content/documents/AlonsoMora-Ondemandhighcapacityridesharing-2017.pdf, Page 2]
This content downloaded from 
64.28.139.201 on Sat, 21 Mar 2026 17:34:57 UTC 
All use subject to https://about.jstor.org/terms


In [ ]:
RAG_PROMPT = """You are a helpful assistant that answers questions
based on the provided context. Follow these rules:
1. Only answer based on the provided context.
2. If the context does not contain enough information, say so.
3. Cite your sources using [Source N] notation.
4. Be concise and accurate.
Context:
{context}
Question: {question}
Answer:"""


In [ ]:
from openai import OpenAI
try:
  client = OpenAI(
  base_url="",
  api_key="",
  )
  def ask_rag(question, vectorstore, k=4):


    retriever = vectorstore.as_retriever(search_kwargs={"k": k})
    docs = retriever.invoke(question)

    context = format_context(docs)

    prompt = RAG_PROMPT.format(context=context, question=question)

    response = client.chat.completions.create(
    model="llama3.3-70b-instruct",
    messages=[
    {"role": "system", "content": "You are a helpful assistant."},
    {"role": "user", "content": prompt}
    ],
    max_tokens=200,
    temperature=0.2
    )
    answer = response.choices[0].message.content
    return answer, docs
    print("ask_rag() defined.")
except Exception as e:
    print("function ask_rag failed:",e)

In [210]:

questions = [
  "How does ride-sourcing services affect urban congestion?",
  "How does the NYC Taxi and Limousine Commission (TLC) regulate driver licensing requirements?",
  "What policies exist to reduce traffic congestion caused by taxis in NYC?",
  "What is the purpose of the congestion relief zone toll",
  "what is TLC Vision Zero"
  ]


In [ ]:
try:
  def ask_and_cite(question, vectorstore, k=4):

    answer, docs = ask_rag(question, vectorstore, k=k)
    print(f"Question: {question}")
    print(f"\nAnswer: {answer}")
    print(f"\n--- Sources ---")
    for i, doc in enumerate(docs, 1):
      source = doc.metadata.get("source", "Unknown")
      page = doc.metadata.get("page", "?")
      print(f"\n[Source {i}]: {source}, Page {page}")
      print(f"Excerpt: {doc.page_content[:200]}...")
      return answer, docs
except Exception as e:
    print("function ask_and_cite failed:",e)

In [ ]:
try:
  questions = [
  "How does ride-sourcing services affect urban congestion?",
  "How does the NYC Taxi and Limousine Commission (TLC) regulate driver licensing requirements?",
  "What policies exist to reduce traffic congestion caused by taxis in NYC?",
  "What is the purpose of the congestion relief zone toll",
  "what is TLC Vision Zero"
  ]


  for q in questions:
    print(f"\nQ: {q}")
    print("-" * 60)
    answer, sources = ask_and_cite(q, vectorstore)
    print(f"A: {answer}")
    print(f"\nSources used: {len(sources)} chunks")
except Exception as e:
    print("pipline testing failed:",e)


Q: How does ride-sourcing services affect urban congestion?
------------------------------------------------------------
Question: How does ride-sourcing services affect urban congestion?

Answer: According to [Source 1], it is not clear whether ride-sourcing is beneficial or unfavorable for traffic congestion. However, if ride-sourcing trips substitute private vehicles or taxis, they should have a secondary influence on congestion, but if they compete with public transportation modes, the impact on congestion is uncertain [Source 1].

--- Sources ---

[Source 1]: /content/documents/2007.00980v1.pdf, Page 1
Excerpt: vital importance in the development of shared transportation for the near future (Narayanan et al.,
2020). Another point of concern is the surge pricing models used, which may considerably increase
th...
A: According to [Source 1], it is not clear whether ride-sourcing is beneficial or unfavorable for traffic congestion. However, if ride-sourcing trips substitute private v

### Task 2.4: RAG Evaluation & Analysis

evaluating the performance of the RAG system using a manually created test set of 10 question-answer pairs.

For each query, we assess:
- Retrieval quality (correct document in top-k results)
- Answer accuracy (factually correct and grounded)

We compute an overall accuracy metric and perform error analysis.

Failures are categorized as:
- Retrieval failures (wrong documents retrieved)
- Generation failures (incorrect answers despite correct context)

We also propose improvements such as better chunking, reranking, or prompt tuning.

In [ ]:
try:
  questions2 = [
  "how many taxis with a capacity of four would be needed to serve 98% of the current taxi demand in NYC, and what is the current fleet size being compared",
  "how do the tip percentages of 'tourists who are theatergoers' compare to those of locals in monetary terms, and what is the reported average tip?",
  "what must happen for ride-sourcing to become a sustainable service?",
  "What is Green Rides Initiative",
  "what is the specific reduction in CO2 emissions if 50% of drivers act as passengers in a scenario with 600 users?",
  "What is the congestion relief zone toll for NYC trips?",
  "Why were shared trips more affected?",
  "what operational changes are required for ride-sourcing to become a sustainable service in terms of emissions and VKT (Vehicle Kilometers Traveled)?",
  "How does the tipping behavior of tourists in NYC taxis impact the geographic allocation of cabs, and what is the potential negative outcome for local residents?",
  "What is the Medallion Relief Program (MRP) successfully"
  ]


  for q in questions2:
    print(f"\nQ: {q}")
    print("-" * 60)
    answer, sources = ask_and_cite(q, vectorstore)
    print(f"A: {answer}")
    print(f"\nSources used: {len(sources)} chunks")
except Exception as e:
    print("rag testing failed:",e)


Q: how many taxis with a capacity of four would be needed to serve 98% of the current taxi demand in NYC, and what is the current fleet size being compared
------------------------------------------------------------
Question: how many taxis with a capacity of four would be needed to serve 98% of the current taxi demand in NYC, and what is the current fleet size being compared

Answer: According to [Source 1], 3,000 taxis with a capacity of four could serve the current taxi rides. However, the context does not provide information on serving 98% of the current demand. The current fleet size being compared is 13,000 taxis [Source 1].

--- Sources ---

[Source 1]: /content/documents/AlonsoMora-Ondemandhighcapacityridesharing-2017.pdf, Page 7
Excerpt: of the taxi rides currently served by over 13,000 taxis could be
served with just 3,000 taxis of capacity four. We observe that a
vehicle capacity of two is sufficient for ride-sharing when a small
tri...
A: According to [Source 1], 3,000 tax


#### 1. Evaluation Results Summary
The table below summarizes the comparison between the system's generated answers and the expected ground-truth data.

for question 1 how many taxis with a capacity of four would be needed to serve 98% of the current taxi demand in NYC, and what is the current fleet size being compared retrieved the correct source AlonsoMora-Ondemandhighcapacityridesharing-2017.pdf and produced a consistent answer

for question 2  how do the tip percentages of 'tourists who are theatergoers' compare to those of locals in monetary terms, and what is the reported average tip? the rag system retrieved the correct source EBSCO-FullText-03_21_2026(2).pdf and produced a consistent answer

for question 3 what must happen for ride-sourcing to become a sustainable service? the rag system retrieved the correct source 2007.00980v1.pdf but failed to produced an answer resulting in not enough information

Following the format provided, here is the completion of the evaluation for the remaining test queries:

for question 4 What is Green Rides Initiative the RAG system retrieved the correct source annual_report_2024.pdf and produced a consistent answer

for question 5 "what is the specific reduction in CO2 emissions the RAG system retrieved the correct source EBSCO-FullText-03_21_2026.pdf and produced a consistent answer

for question 6 What is the congestion relief zone toll the RAG system retrieved the correct source 155761-the-effect-of-the-new-york-city-congestion-toll...pdf but failed to produce an answer, correctly noting that the document discusses effects rather than specific toll amounts.

for question 7 Why were shared trips more affected? the RAG system retrieved the correct source 2007.00980v1.pdf and produced a consistent answer

for question 8 what operational changes are required..., the RAG system retrieved the correct source 2007.00980v1.pdf but produced a partially limited answer, correctly identifying that the source mentions a necessary change in modus operandi but lacks specific details.

for question 9 How does tipping behavior impact geographic allocationthe RAG system retrieved the correct source EBSCO-FullText-03_21_2026(2).pdf and produced a consistent answer

for question 10 What is the Medallion Relief Program (MRP) successfully, the RAG system retrieved the correct source annual_report_2024.pdf and produced a consistent answer

## 2. Accuracy Metric
for a total of 10 question asked, there was a total 8 correct answers with 10 correct document sources, question 3 failed to answer and question eight produced a partial answer.resulting in a 100% correct retirval but a 80% correct answer

**Overall Accuracy: 80%**

#### 3. Error Analysis
* **Retrieval Performance:** The retriever performed well with the 1000-character chunk size. as from testing with academic paper by external sources it was determinded that 1000 chucks was a sweetspot as it suited the sciencetific concise paragrph style

* **Observed Weaknesses:** the system retrieved correct document and chunks but failed in inferring and interperting information compared to purely retrival

###Part 3: Integrated Analytics Application


### Task 3.1: Query Router

This component classifies natural language queries into:
- DATA (Spark)
- DOCUMENT (RAG)
- HYBRID (both)

We use an LLM to generate structured JSON output containing:
- Predicted category
- Explanation

A test set of 15 queries is used to evaluate classification accuracy.

Ambiguous queries default to HYBRID to ensure comprehensive handling.

In [ ]:
import json
ROUTER_PROMPT = """
  You are a query classifier for a data analytics system.

  Classify the user's question into ONE of these categories:

  1. DATA → Answerable using structured taxi trip data (e.g., averages, counts, trends)
  2. DOCUMENT → Answerable using document knowledge (e.g., policies, rules, explanations)
  3. HYBRID → Requires BOTH data and documents

  Return ONLY valid JSON:

  {
    "category": "DATA | DOCUMENT | HYBRID",
    "reasoning": "short explanation"
  }

  Rules:
  - If unsure, choose HYBRID
  - DATA = numbers, aggregations, trends
  - DOCUMENT = definitions, rules, explanations
  - HYBRID = comparison between real data and policies
  """


In [195]:
def route_query(question):
      response = client.chat.completions.create(
          model="llama3.3-70b-instruct",
          messages=[
              {"role": "system", "content": ROUTER_PROMPT},
              {"role": "user", "content": question}
          ],
          max_tokens=200,
          temperature=0.2
      )
      output = response.choices[0].message.content
      try:
          result = json.loads(output)
      except:
          result = {
              "category": "HYBRID",
              "reasoning": "undecided error"
          }
      return result


In [206]:
try:
    questions2 = [
        {"query": "What was the average fare amount for trips in January 2024?", "expected": "DATA"},
        {"query": "Show me the top 5 pickup locations by trip count.", "expected": "DATA"},
        {"query": "What is the total congestion surcharge collected on Mondays?", "expected": "DATA"},
        {"query": "Compare the average trip distance between Vendor 1 and Vendor 2.", "expected": "DATA"},
        {"query": "Which hour of the day has the highest frequency of credit card payments?", "expected": "DATA"},
        {"query": "What are the specific requirements of the Green Rides Initiative?", "expected": "DOCUMENT"},
        {"query": "According to the TLC, how is 'on-time performance' defined for buses?", "expected": "DOCUMENT"},
        {"query": "What are the driver safety regulations for for-hire vehicles?", "expected": "DOCUMENT"},
        {"query": "What does the Medallion Relief Program provide to owners?", "expected": "DOCUMENT"},
        {"query": "What are the three main policy implications of the tipping study?", "expected": "DOCUMENT"},
        {"query": "Are actual January tip percentages higher than the averages cited in the tourism study?", "expected": "HYBRID"},
        {"query": "How does the current fleet's EV adoption rate compare to the 2030 Green Rides goals?", "expected": "HYBRID"},
        {"query": "Do trips in the congestion zone show a decrease in VKT as recommended by the sustainability paper?", "expected": "HYBRID"},
        {"query": "Based on the 3,000-taxi optimization study, how many of our current trips would be served?", "expected": "HYBRID"},
        {"query": "Is the average driver income from the dataset sufficient according to the TLC living wage policy?", "expected": "HYBRID"}
    ]
    correct_count = 0

    print(f"{'query'} | {'sredicted'} | {'reasoning'} |{'status'}")
    for item in questions2:
        prediction = route_query(item['query'])

        category= prediction.get('category', 'DATA')
        if category ==item['expected']:
          correct_count += 1
          status = "correct"
        else :
          status = "incorrect"
        reasoning  =prediction.get('reasoning','explaination')
        print(f"{item['query'] } -- {category} -- {reasoning} -- {status}\n")

    accuracy = (correct_count /len(questions2))*100
    print("Total queries", len(questions2))
    print("Correct classification", correct_count)
    print("Final Router Accuracy",  accuracy)
except Exception as e:
    print("Query routing testing failed:", e)

query | sredicted | reasoning |status
What was the average fare amount for trips in January 2024? -- DATA -- The question asks for a numerical value (average fare amount) that can be calculated from structured taxi trip data. -- correct

Show me the top 5 pickup locations by trip count. -- HYBRID -- undecided error -- incorrect

What is the total congestion surcharge collected on Mondays? -- DATA -- The question asks for a specific total amount collected, which can be calculated using structured taxi trip data. -- correct

Compare the average trip distance between Vendor 1 and Vendor 2. -- DATA -- The question asks for a comparison of average trip distances, which can be calculated using structured taxi trip data. -- correct

Which hour of the day has the highest frequency of credit card payments? -- DATA -- This question requires analyzing structured payment data to determine the hour with the highest frequency of credit card payments, which involves numerical aggregation. -- correct


the query router correctly predicted 13 out of 15 quires with a accuracy of 86%. it correctly predicted 100%  of all documents based quires but in the data quries if failed to decide if a query is a data but defaulted to hybrid. and for the question 14, the addition of specfic values made the model halucinate thinking it was a data instead of making a insight.

### Task 3.2: Data Query Handler

This module translates natural language queries into Spark SQL.

Pipeline:
1. Convert question → SQL query (LLM)
2. Execute query in Spark
3. Convert results → natural language answer

We include error handling by retrying failed queries with additional context.

This enables non-technical users to interact with large datasets using plain English.

In [199]:

SQL_PROMPT = """
You are a Spark SQL expert.CRITICAL RULE: Always handle division by zero.

Convert the user's question into a valid Spark SQL query.

Table name: data

Columns include:
- pickup_hour
- pickup_day
- fare_amount
- tip_amount
- total_amount
- trip_distance

Rules:
- Return ONLY SQL (no explanation)
- Use correct Spark SQL syntax
- Use aggregation functions when needed (AVG, COUNT, etc.)

Example:
Q: What is the average fare?
A: SELECT AVG(fare_amount) AS avg_fare FROM trips
"""

In [200]:
try:
  def generate_sql(question):
      response = client.chat.completions.create(
          model="llama3.3-70b-instruct",
          messages=[
              {"role": "system", "content": SQL_PROMPT},
              {"role": "user", "content": question}
          ],
          max_tokens=200,
          temperature=0.2
      )

      sql = response.choices[0].message.content.strip()
      if "```" in sql:
          sql = sql.replace("```sql", "").replace("```", "").strip()
      return sql
except Exception as e:
    print("generate_sql failed:",e)

try:
  def run_sql(sql):
      result_df = spark.sql(sql)
      return result_df
except Exception as e:
    print(" run_sql failed:",e)

In [201]:
ANSWER_PROMPT = """
You are a data analyst.

Given:
1. A user question
2. SQL query result

Write a clear natural language answer.

Be concise and interpret the result.

Question: {question}
Result: {result}
Answer:
"""

In [202]:
def generate_answer(question, result):
    response = client.chat.completions.create(
        model="llama3.3-70b-instruct",
        messages=[
            {"role": "user", "content": ANSWER_PROMPT.format(
                question=question,
                result=result
            )}
        ],
        max_tokens=200,
        temperature=0.2
    )

    return response.choices[0].message.content

In [203]:
def ask_data(question):

    #convert question into sql
    sql = generate_sql(question)
    print("Generated sql")
    #execute sql
    sql_result = run_sql(sql)
    result = sql_result.show(5)

    #convert results back into a natural language
    rows = sql_result.limit(5).collect()
    result_str = [row.asDict() for row in rows]

    answer = generate_answer(question, result_str)

    return answer

In [204]:
try:
  data_questions = [
      "What is the average fare?",
      "How many trips are there?",
      "What is the average tip amount?",
      "What is the busiest pickup hour?",
      "What is the average trip distance?"
  ]

  for q in data_questions:
     ans=  ask_data(q)
     print(ans)
except Exception as e:
    print(" sql  query handler failed:",e)

Generated sql
+------------------+
|          avg_fare|
+------------------+
|18.484269959662264|
+------------------+

The average fare is $18.48.
Generated sql
+-----------+
|total_trips|
+-----------+
|    2870540|
+-----------+

There are 2,870,540 trips.
Generated sql
+------------------+
|           avg_tip|
+------------------+
|3.4010956161563044|
+------------------+

The average tip amount is approximately $3.40.
Generated sql
+-----------+---------+
|pickup_hour|num_trips|
+-----------+---------+
|         18|   206346|
+-----------+---------+

The busiest pickup hour is 6 PM (18:00), with approximately 206,346 trips.
Generated sql
+-----------------+
|avg_trip_distance|
+-----------------+
|3.731905766859004|
+-----------------+

The average trip distance is approximately 3.73 miles.


the sql query handler corrected output sql as well as converting the sql results into a natural language. however a limitation of testing is the test set was composed of simple quries but these output results would not be indicative for more complex and business oriented data. the natural language output while moderate quality for these test quries, more informative explaination will be required if to be used outside academics however due to no testing this is to be determined

### Task 3.3: End-to-End System Demonstration

In this section, we demonstrate the full system workflow.

Each query goes through:
1. Classification
2. Routing
3. Processing (Spark, RAG, or both)
4. Final response generation

For HYBRID queries, outputs from both systems are combined.

This demonstrates a real-world architecture where structured and unstructured data sources are integrated.

In [205]:
def process_query(question):
    print(f"\n{'='*60}\nUSER QUERY: {question}")
    route = route_query(question)
    cat = route['category']
    print(" qurey catagory:", cat,"with reasoning being ",route['reasoning'])

    if cat == "DATA":
        ans = ask_data(question)
        print("answer ", ans)

    if cat == "DOCUMENT":
        ans,_ = ask_and_cite(question, vectorstore)

    if cat == "HYBRID":
        data_ans = ask_data(question)
        doc_ans, _ = ask_and_cite(question, vectorstore)
        hybrid_prompt = f"""gnerate a single answer using both of these insights.
        Data Insight (from Spark): {data_ans}
        Document Insight (from RAG): {doc_ans}"""

        response = client.chat.completions.create(
            model="llama3.3-70b-instruct",
            messages=[{"role": "user", "content": hybrid_prompt}],
            max_tokens=200,
            temperature=0.2
        )
        print( "ans ", response.choices[0].message.content)

final_test_queries = [
    "What is the average trip distance for January 2024?",
    "Explain the requirements for the Green Rides Initiative.",
    "How do actual average tips compare to the 'tourist theatergoer' study findings?",
    "What are the CO2 reduction benefits of 50% driver flexibility?",
    "Which hour of the day has the highest total fare amount?"
]

for i, query in enumerate(final_test_queries):
    print(f"query {i+1}: {query}")
    try:
        response = process_query(query)
        print(f"result: {response}")
    except Exception as e:
        print(f"system Error: {e}")


query 1: What is the average trip distance for January 2024?

USER QUERY: What is the average trip distance for January 2024?
 qurey catagory: DATA with reasoning being  The question asks for a numerical value (average trip distance) that can be calculated from structured taxi trip data.


{"ts": "2026-03-31 15:04:11.838", "level": "ERROR", "logger": "SQLQueryContextLogger", "msg": "[UNRESOLVED_COLUMN.WITH_SUGGESTION] A column, variable, or function parameter with name `pickup_day` cannot be resolved. Did you mean one of the following? [`pickup_hour`, `extra`, `mta_tax`, `VendorID`, `pickup_day_of_week`]. SQLSTATE: 42703", "context": {"errorClass": "UNRESOLVED_COLUMN.WITH_SUGGESTION"}, "exception": {"class": "Py4JJavaError", "msg": "An error occurred while calling o36.sql.\n: org.apache.spark.sql.AnalysisException: [UNRESOLVED_COLUMN.WITH_SUGGESTION] A column, variable, or function parameter with name `pickup_day` cannot be resolved. Did you mean one of the following? [`pickup_hour`, `extra`, `mta_tax`, `VendorID`, `pickup_day_of_week`]. SQLSTATE: 42703; line 3 pos 6;\n'Project ['AVG('trip_distance) AS avg_trip_distance#35640]\n+- 'Filter (('pickup_day >= 2024-01-01) AND ('pickup_day < 2024-02-01))\n   +- SubqueryAlias data\n      +- View (`data`, [VendorID#24765, tpep_p

Generated sql
system Error: [UNRESOLVED_COLUMN.WITH_SUGGESTION] A column, variable, or function parameter with name `pickup_day` cannot be resolved. Did you mean one of the following? [`pickup_hour`, `extra`, `mta_tax`, `VendorID`, `pickup_day_of_week`]. SQLSTATE: 42703; line 3 pos 6;
'Project ['AVG('trip_distance) AS avg_trip_distance#35640]
+- 'Filter (('pickup_day >= 2024-01-01) AND ('pickup_day < 2024-02-01))
   +- SubqueryAlias data
      +- View (`data`, [VendorID#24765, tpep_pickup_datetime#24766, tpep_dropoff_datetime#24767, passenger_count#24768L, trip_distance#24769, RatecodeID#24770L, store_and_fwd_flag#24771, PULocationID#24772, DOLocationID#24773, payment_type#24774L, fare_amount#24775, extra#24776, mta_tax#24777, tip_amount#24778, tolls_amount#24779, improvement_surcharge#24780, total_amount#24781, congestion_surcharge#24782, Airport_fee#24783, trip_duration_minutes#25080, trip_speed_mph#25081, pickup_hour#25082, pickup_day_of_week#25083, tip_percentage#25084])
         +

**reflection**

Strengths and Performance:
the accuracy query router  achieved a classification success of 86% but to do the gracefully failing but setting an undecided to hybird conpensates for any failure. The integration of spark distributed processing allows for sub-second responses on aggregations of millions of rows, while the RAG pipeline provides deep policy context that structured data alone cannot offer. The implementation of a self-correction loop in the sql handler further enhances robustness, as demonstrated during query 1 where the system successfully recovered from a schema hallucination by identifying the correct tpep_pickup_datetime field after an initial failure.

limitations:
Retrieval failures occurred when relevant information was not captured in the top-k chunks especially for more complex queries. On the generation side, the LLM occasionally produced  slightly hallucinated answers when context was insufficient.

Future Work and Scalability:
improvements could include better retrieval strategies such as reranking enhanced prompt engineering, and more robust error handling for SQL generation.

## AI Tools Used

AI tools such as ChatGPT were used to assist with:
- Structuring code and workflow
- Debugging and refining logic
- Improving documentation and explanations

All outputs were carefully reviewed and modified to ensure correctness and understanding.

This submission reflects my own work and understanding.